### ISSUE TITLE
Domination 1914 "bunkers" technology broken

### ISSUE THREAD

**Issue by samstreeter** [triplea-game/triplea#6861]
The Domination 1914 map has a technology known as "bunkers" that raises the defense of the trench unit from 0 to 1. The bug is that this defense value is now acting as a first strike when it should not.

### If playing a prerelease, which version are you using?

Game Version:

### If playing a prerelease, does this happen on the latest release?

- [ ] yes
- [ ] no

**Comment by trevan** [comment]
The "bunkers" technology is working correctly.  The problem is that the trench is an `isSuicide` unit and `Defending Suicide and Munition Units Do Not Fire` is true.

In 1.9, this basically meant that the trench would do nothing until the "bunkers" technology was added and then it could do something on the defense.

In 2.0, the `isSuicide` was split apart and now all `isSuicide` units gain `isFirstStrike`.  So when the trench has a bunker tech, it can defend and it will defend with `isFirstStrike`.

The only reason why trench is `isSuicide` is so that the the gas unit (which is `isSuicide` and has attack) can't target it.  It doesn't actually utilize any suicide ability.

Some `isSuicide` units should have `isFirstStrike` on defense.  This is when `Defending Suicide and Munition Units Do Not Fire` is false.

**Comment by Cernelius** [comment]
> The "bunkers" technology is working correctly. The problem is that the trench is an `isSuicide` unit and `Defending Suicide and Munition Units Do Not Fire` is true.

It is not. Obviously whoever has the ownership of this repository can anytime decide how anything works, but, aside from that, the simple fact that it is not working as it always did necessarily implies that the current behaviour is not correct, unless it can be demostrated that the previous behaviour was wrong, thus the behavioural change is a problem fix (in that case, how it worked would be wrong). Either how it worked is wrong or how it works is wrong, if they are not both wrong. They cannot be both right. In my opinion, the wording of the property itself doesn't matter (properties and other items are often poorly worded, but only the behaviour matters, which is in turn either based on intended rules, not on naming, or on nothing (if there are no referring rules, thus the behaviour is based on nothing, a mapmaker can really refer to nothing else but the behaviour itself)). What I'm trying to say is that a mapmaker made this game and made it work under some rules, thus unless you demonstrate that these rules are wrong (thus the game had always worked wrongly until this new behaviour has applied), the existing behaviour at the time the game was made can only be assumed as right, can't it? If not, what is left to mapmakers to follow, when they make anything else but "cloning" a game with a full rulebook behind (which I believe is not the case here)?

> In 1.9, this basically meant that the trench would do nothing until the "bunkers" technology was added and then it could do something on the defense.

I would rather say that in both cases this means that the trench would defend normally, albeit with defence power 0 with no tech and defence power 1 with tech. Obviously, you can say that defending at power 0 is doing nothing, which in my opinion is not correct, as if you can get taken as casualty, then you are already doing something on defence. Really, the difference between a tech non-improved trench and a tech improved one is about the same difference that there is in attack between a conscript and an infantry. I don't think many will follow you if you say that "a conscript can do nothing on the offence, while an infantry can do something".

> In 2.0, the `isSuicide` was split apart and now all `isSuicide` units gain `isFirstStrike`. So when the trench has a bunker tech, it can defend and it will defend with `isFirstStrike`.

Since what that property always did was to impede the unit to have both behaviours when on the defence, it seems logical to me to affirm that the fact that it is now failing fully to assure this is a problem, unless this change is wanted by the program owner (whoever it is) or it can be demostrated that `isSuicide` units should have always first striked when they were not suiciding (which seems to me this is not the case, as that property was always meant to have suicide units acting normally in defence, which also means not having first strike (unless they are specifically meant to be first strike units when non suciding)). In my opinion, the unit in question should not first strike on defence unless the first strike option is specifically set true for the unit in addition to the suicide one (I don't actually know if this would work, but this is the direction I suggest the fix possibly to go).

> The only reason why trench is `isSuicide` is so that the the gas unit (which is `isSuicide` and has attack) can't target it. It doesn't actually utilize any suicide ability.

It is not the only reason, as you can test that, for example, if you attack with some infantries and some trenches a territory defended by conscripts, you will lose all trenches (if the trenches would not be suicide units, you would be able to take them as casualties, instead of the infantries, and you would also avoid losing them all for nothing). Of course, this, in turn, can only happen if those conscripts attacked alone the territory defended by trenches alone (stalemate), which is in turn an illegal move, as in v3 offenceless units can never attack alone, but one that, apparently, the players of this map like to do (even if it is substantially exploiting a problem with the program). By the way, one could argue that suicide units should not suicide if they have power 0, but this is another matter entirely (we should open another issue for it).

> Some `isSuicide` units should have `isFirstStrike` on defense. This is when `Defending Suicide and Munition Units Do Not Fire` is false.

Correct, but, again, if `isFirstStrike` is gained because of having `isSuicide`, then I believe that when `Defending Suicide and Munition Units Do Not Fire` is true units should lose both the suicide and the first strike behaviours on defence, unless the units are specifically set as `isFirstStrike`. This is mostly for consistency on how the property has always worked, thus on how it can only be assumed it is supposed to work, unless someone can demostrate the behaviour has always been wrong.

**Comment by trevan** [comment]
> > The "bunkers" technology is working correctly. The problem is that the trench is an `isSuicide` unit and `Defending Suicide and Munition Units Do Not Fire` is true.
> 
> It is not. Obviously whoever has the ownership of this repository can anytime decide how anything works, but, aside from that, the simple fact that it is not working as it always did necessarily implies that the current behaviour is not correct, unless it can be demostrated that the previous behaviour was wrong, thus the behavioural change is a problem fix (in that case, how it worked would be wrong). Either how it worked is wrong or how it works is wrong, if they are not both wrong. They cannot be both right. In my opinion, the wording of the property itself doesn't matter (properties and other items are often poorly worded, but only the behaviour matters, which is in turn either based on intended rules, not on naming, or on nothing (if there are no referring rules, thus the behaviour is based on nothing, a mapmaker can really refer to nothing else but the behaviour itself)). What I'm trying to say is that a mapmaker made this game and made it work under some rules, thus unless you demonstrate that these rules are wrong (thus the game had always worked wrongly until this new behaviour has applied), the existing behaviour at the time the game was made can only be assumed as right, can't it? If not, what is left to mapmakers to follow, when they make anything else but "cloning" a game with a full rulebook behind (which I believe is not the case here)?
> 
> > In 1.9, this basically meant that the trench would do nothing until the "bunkers" technology was added and then it could do something on the defense.
> 
> I would rather say that in both cases this means that the trench would defend normally, albeit with defence power 0 with no tech and defence power 1 with tech. Obviously, you can say that defending at power 0 is doing nothing, which in my opinion is not correct, as if you can get taken as casualty, then you are already doing something on defence. Really, the difference between a tech non-improved trench and a tech improved one is about the same difference that there is in attack between a conscript and an infantry. I don't think many will follow you if you say that "a conscript can do nothing on the offence, while an infantry can do something".
> 
> > In 2.0, the `isSuicide` was split apart and now all `isSuicide` units gain `isFirstStrike`. So when the trench has a bunker tech, it can defend and it will defend with `isFirstStrike`.
> 
> Since what that property always did was to impede the unit to have both behaviours when on the defence, it seems logical to me to affirm that the fact that it is now failing fully to assure this is a problem, unless this change is wanted by the program owner (whoever it is) or it can be demostrated that `isSuicide` units should have always first striked when they were not suiciding (which seems to me this is not the case, as that property was always meant to have suicide units acting normally in defence, which also means not having first strike (unless they are specifically meant to be first strike units when non suciding)). In my opinion, the unit in question should not first strike on defence unless the first strike option is specifically set true for the unit in addition to the suicide one (I don't actually know if this would work, but this is the direction I suggest the fix possibly to go).
> 
> > The only reason why trench is `isSuicide` is so that the the gas unit (which is `isSuicide` and has attack) can't target it. It doesn't actually utilize any suicide ability.
> 
> It is not the only reason, as you can test that, for example, if you attack with some infantries and some trenches a territory defended by conscripts, you will lose all trenches (if the trenches would not be suicide units, you would be able to take them as casualties, instead of the infantries, and you would also avoid losing them all for nothing). Of course, this, in turn, can only happen if those conscripts attacked alone the territory defended by trenches alone (stalemate), which is in turn an illegal move, as in v3 offenceless units can never attack alone, but one that, apparently, the players of this map like to do (even if it is substantially exploiting a problem with the program). By the way, one could argue that suicide units should not suicide if they have power 0, but this is another matter entirely (we should open another issue for it).
> 
> > Some `isSuicide` units should have `isFirstStrike` on defense. This is when `Defending Suicide and Munition Units Do Not Fire` is false.
> 
> Correct, but, again, if `isFirstStrike` is gained because of having `isSuicide`, then I believe that when `Defending Suicide and Munition Units Do Not Fire` is true units should lose both the suicide and the first strike behaviours on defence, unless the units are specifically set as `isFirstStrike`. This is mostly for consistency on how the property has always worked, thus on how it can only be assumed it is supposed to work, unless someone can demostrate the behaviour has always been wrong.

I'm sorry, I meant that the "bunkers" technology is correctly being applied and the unit is correctly getting the defense value.  It isn't the "bunkers" technology that is creating the issue.  I agree that there is a bug, I was just clarifying what was the actual issue.  The issue is caused when a unit is an `isSuicide` unit and `Defending Suicide and Munition Units Do Not Fire` is true.  That unit shouldn't have first strike on defense.

Can you explain the example you gave?  I'm having difficulty replicating it.

**Comment by ron-murhammer** [comment]
@trevan Would a better fix to this be to remove IsSuicide from bunkers and just make gas units not be able to target them?

**Comment by trevan** [comment]
@ron-murhammer I believe that would fix it for this map (though @Cernelius has indicated that there is another reason for the isSuicide that I'm trying to replicate).

But other maps could have similar issues.  Any map with an `isSuicide` unit and `Defending Suicide and Munition Units Do Not Fire` is true will make that unit have a first strike defense hit.  Previously, those units were not a part of the special suicide defense phase.

**Comment by Cernelius** [comment]
> @trevan Would a better fix to this be to remove IsSuicide from bunkers and just make gas units not be able to target them?

In this case, I suppose you should also take into consideration what happens when trenches and enemy conscripts are in the same zone. Currently the suicide behaviour substantially mostly means that the conscripts destroy the trenches, albeit you have to wait the trenches' owner turn for that to happen (unless an ally clears the zone beforehand).

@trevan I'm not sure what is not clear to you. Maybe the confusion derives from the fact that I added some infantries to the fray. You don't need to add them, or anything else, really. Just trenches (alone or not) attacking conscripts (alone or not, but in practice can only be alone) will suicide (of course, since the trenches cannot move, this can happen only after the conscripts moved in themselves).

**Comment by Cernelius** [comment]
> @trevan Would a better fix to this be to remove IsSuicide from bunkers and just make gas units not be able to target them?

I'm not sure if by this you mean that you would leave in place the behaviour that suicide units that don't suicide on defence will first strike, instead. I guess you can do that, but, if so, please document it clearly in pos2 (where the property `Defending Suicide and Munition Units Do Not Fire` is). Obviously, the property itself is terribly worded anyway, as the behaviour is that they fire, just not in a suicidal way. What the property is literally saying is that suicide units will also be set at defence 0, but I believe this has never been the case.

**Comment by trevan** [comment]
> @trevan I'm not sure what is not clear to you. Maybe the confusion derives from the fact that I added some infantries to the fray. You don't need to add them, or anything else, really. Just trenches (alone or not) attacking conscripts (alone or not, but in practice can only be alone) will suicide (of course, since the trenches cannot move, this can happen only after the conscripts moved in themselves).

It was that last part that wasn't clear.  I didn't think that the trenches could attack the conscripts because the trenches have no movement.  So I was confused on what you were describing.  Thanks.

**Comment by Cernelius** [comment]
> > @trevan I'm not sure what is not clear to you. Maybe the confusion derives from the fact that I added some infantries to the fray. You don't need to add them, or anything else, really. Just trenches (alone or not) attacking conscripts (alone or not, but in practice can only be alone) will suicide (of course, since the trenches cannot move, this can happen only after the conscripts moved in themselves).
> 
> It was that last part that wasn't clear. I didn't think that the trenches could attack the conscripts because the trenches have no movement. So I was confused on what you were describing. Thanks.

I would actually argue that you may be right, since conscripts alone should never attack anything. However, while conscripts alone attacking defended territories is surely an illegal move (the players of this map usually do just because they don't know the rules), I would argue that conscripts alone attacking either empty territories or territories with only trenches in them might be a permissible move (and one that I would permit, were I to rule over it).

**Comment by Cernelius** [comment]
There would be also the case in which you are left with conscripts alone attacking because the offenceful units were to be offloaded from transports that were sunk or retreated, but I don't want to go too hard off topic. Suffice it to say, it is possible for trenches to end up, on the turn of their controller, to be in the same territory as enemy conscripts, one way or the other (some more or less legal than others).
